### PR TITLE
Translate reconciliation

### PR DESCRIPTION
**Pull request by codejet**

### CODE REVIEW

File: content/docs/reconciliation.md [hunk 1]
```diff
@@ -1,37 +1,37 @@
 ---
 id: reconciliation
-title: Reconciliation
+title: Abgleich
 permalink: docs/reconciliation.html
 ---
 
-React provides a declarative API so that you don't have to worry about exactly what changes on every update. This makes writing applications a lot easier, but it might not be obvious how this is implemented within React. This article explains the choices we made in React's "diffing" algorithm so that component updates are predictable while being fast enough for high-performance apps.
+React bietet eine deklarative API, sodass du dich nicht bei jeder Aktualisierung um die genauen Änderungen kümmern musst. Dies erleichtert das Erstellen von Anwendungen erheblich. Es ist jedoch möglicherweise nicht offensichtlich, wie dies in React implementiert ist. Dieser Artikel erklärt welche Entscheidungen wir in Reacts Vergleichsalgorithmus getroffenen haben, damit Komponentenaktualisierungen vorhersehbar und gleichzeitig schnell genug für leistungsstarke Apps sind.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
React bietet eine deklarative API, sodass du dich nicht bei jeder Aktualisierung um die genauen Änderungen kümmern musst. Dies erleichtert das Erstellen von Anwendungen erheblich. Es ist jedoch möglicherweise nicht offensichtlich, wie dies in React implementiert ist. Dieser Artikel erklärt welche Entscheidungen wir in Reacts Vergleichsalgorithmus getroffen haben, damit Komponentenaktualisierungen vorhersehbar und gleichzeitig schnell genug für leistungsstarke Apps sind.
```

File: content/docs/reconciliation.md [hunk 2]
```diff
@@ -1,37 +1,37 @@
...
+Wenn du React verwendest, kannst du dir die `render()`-Funktion so vorstellen, dass sie zu einem bestimmten Zeitpunkt einen Baum von React-Elementen erstellt. Bei der nächsten Aktualisierung von State oder Props gibt diese `render()`-Funktion einen anderen Baum von React-Elementen zurück. React muss dann herausfinden, wie die Benutzeroberfläche effizient aktualisiert werden kann, um sie an den neuesten Baum anzupassen.
 
-There are some generic solutions to this algorithmic problem of generating the minimum number of operations to transform one tree into another. However, the [state of the art algorithms](https://grfia.dlsi.ua.es/ml/algorithms/references/editsurvey_bille.pdf) have a complexity in the order of O(n<sup>3</sup>) where n is the number of elements in the tree.
+Es gibt einige generische Lösungen für dieses algorithmische Problem, die Mindestanzahl von Operationen zu generieren um einen Baum in einen anderen umzuwandeln. Die Algorithmen nach dem aktuellen Stand der Technik weisen jedoch eine Komplexität in der Größenordnung von O(n3) auf, wobei n die Anzahl der Elemente im Baum ist.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
Es gibt einige generische Lösungen für dieses algorithmische Problem, die Mindestanzahl von Operationen zu generieren um einen Baum in einen anderen umzuwandeln. Die modernsten Algorithmen weisen jedoch eine Komplexität in der Größenordnung von O(n3) auf, wobei n die Anzahl der Elemente im Baum ist.
```

Comment (miku86) [reply content/docs/reconciliation.md]: IMHO "Algorithmen nach dem aktuellen Stand der Technik" sounds very odd to me.

File: content/docs/reconciliation.md [hunk 3]
```diff
@@ -1,37 +1,37 @@
...
 
-If we used this in React, displaying 1000 elements would require in the order of one billion comparisons. This is far too expensive. Instead, React implements a heuristic O(n) algorithm based on two assumptions:
+Würden wir dies in React verwenden, dann wären für das Anzeigen von 1000 Elementen Vergleiche in der Größenordnung von einer Milliarde notwenig, was viel zu aufwendig ist. Stattdessen implementiert React einen heuristischen O(n)-Algorithmus, der auf zwei Annahmen basiert:
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
Würden wir diese in React verwenden, dann wären für das Anzeigen von 1000 Elementen Vergleiche in der Größenordnung von einer Milliarde notwenig, was viel zu aufwendig ist. Stattdessen implementiert React einen heuristischen O(n)-Algorithmus, der auf zwei Annahmen basiert:
```

Comment (miku86) [reply content/docs/reconciliation.md]: As far as I understand, `this` refers to `Lösungen` or `Algorithmen`.

File: content/docs/reconciliation.md [hunk 4]
```diff
@@ -1,37 +1,37 @@
...
 
-Whenever the root elements have different types, React will tear down the old tree and build the new tree from scratch. Going from `<a>` to `<img>`, or from `<Article>` to `<Comment>`, or from `<Button>` to `<div>` - any of those will lead to a full rebuild.
+Wenn die Wurzelelemente unterschiedliche Typen haben, entfernt React den alten Baum und erstellt den neuen Baum von Grund auf neu. Ein Wechsel von `<a>` nach `<img>` oder von `<Article>` nach `<Comment>` oder von `<Button>` nach `<div>` führt zu einer vollständigen Neuerstellung.
 
-When tearing down a tree, old DOM nodes are destroyed. Component instances receive `componentWillUnmount()`. When building up a new tree, new DOM nodes are inserted into the DOM. Component instances receive `componentWillMount()` and then `componentDidMount()`. Any state associated with the old tree is lost.
+Beim Entfernen eines Baumes werden alte DOM-Knoten gelöscht. Komponenteninstanzen erhalten `componentWillUnmount()`. Beim Aufbau eines neuen Baums werden neue DOM-Knoten in das DOM eingefügt. Komponenteninstanzen erhalten `componentWillMount()` und dann `componentDidMount()`. Jeder mit dem alten Baum verknüpfte State geht verloren.
```
Comment (miku86) [inline content/docs/reconciliation.md]: I think in this case `to receive` means `erhalten den Befehl zum Ausführen`.

@ph1p what's your opinion on this?

File: content/docs/reconciliation.md [hunk 5]
```diff
@@ -1,37 +1,37 @@
...
 
-Any components below the root will also get unmounted and have their state destroyed. For example, when diffing:
+Alle Komponenten unterhalb der Wurzel werden ebenfalls unmountet und ihr State wird gelöscht. Zum Beispiel beim Vergleichen von:
```
Comment (miku86) [inline content/docs/reconciliation.md]: I think by using the passive, there needs to be a `ge-`.
`ungemountet`? 
`geunmountet`?

File: content/docs/reconciliation.md [hunk 6]
```diff
@@ -1,37 +1,37 @@
...
 
-```xml
+```js
```
Comment (miku86) [inline content/docs/reconciliation.md]: @gaearon @tesseralis 

I couldn't find a definite answer if `xml` is the correct code format to use.

As far as I understand, you are using `xml` for HTML (e.g. LoC 50) and JSX (e.g LoC 34), am I correct?

File: content/docs/reconciliation.md [hunk 7]
```diff
@@ -41,45 +41,45 @@ Any components below the root will also get unmounted and have their state destr
...
 <div className="before" title="stuff" />
 
 <div className="after" title="stuff" />
 ```
 
-By comparing these two elements, React knows to only modify the `className` on the underlying DOM node.
+Durch den Vergleich dieser beiden Elemente weiß React, dass nur der Klassenname auf dem zugrunde liegenden DOM-Knoten zu ändern ist.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
Durch den Vergleich dieser beiden Elemente weiß React, dass nur der Klassenname (`className`) auf dem zugrunde liegenden DOM-Knoten zu ändern ist.
```

Comment (miku86) [reply content/docs/reconciliation.md]: I think the reader would benefit from emphasizing this reserved "keyword".

File: content/docs/reconciliation.md [hunk 8]
```diff
@@ -128,30 +128,30 @@ In order to solve this issue, React supports a `key` attribute. When children ha
...
 
-[Here](codepen://reconciliation/index-used-as-key) is an example of the issues that can be caused by using indexes as keys on CodePen, and [here](codepen://reconciliation/no-index-used-as-key) is an updated version of the same example showing how not using indexes as keys will fix these reordering, sorting, and prepending issues.
+[Hier](codepen://reconciliation/index-used-as-key) findest du ein Codepent-Beispiel für Probleme, die durch die Verwendung von Indizes als Schlüssel verursacht werden können. [Hier] (codepen://reconciliation/no-index-used-as-key) findest du eine aktualisierte Version desselben Beispiels, in der gezeigt wird, wie durch die Nichtverwendung von Indizes als Schlüssel diese Probleme beim Neuordnen, Sortieren und Voranstellen behoben werden.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
[Hier](codepen://reconciliation/index-used-as-key) findest du ein Codepen-Beispiel für Probleme, die durch die Verwendung von Indizes als Schlüssel verursacht werden können. [Hier] (codepen://reconciliation/no-index-used-as-key) findest du eine aktualisierte Version desselben Beispiels, in der gezeigt wird, wie durch die Nichtverwendung von Indizes als Schlüssel diese Probleme beim Neuordnen, Sortieren und Voranstellen behoben werden.
```

File: content/docs/reconciliation.md [hunk 9]
```diff
@@ -128,30 +128,30 @@ In order to solve this issue, React supports a `key` attribute. When children ha
...
+Wir verfeinern die Heuristiken regelmäßig, um gängige Anwendungsfälle zu beschleunigen. In der aktuellen Implementierung kannst du die Tatsache ausdrücken, dass ein Teilbaum zwischen seinen Geschwister-Elementen verschoben wurde. Du kannst jedoch nicht feststellen, dass er an eine andere Stelle verschoben wurde. Der Algorithmus rendert den vollständigen Teilbaum erneut.
 
-Because React relies on heuristics, if the assumptions behind them are not met, performance will suffer.
+Da sich React auf Heuristiken stützt, wirkt es sich negativ auf die Leistung aus, wenn die zugrunde liegenden Annahmen nicht erfüllt werden.
 
-1. The algorithm will not try to match subtrees of different component types. If you see yourself alternating between two component types with very similar output, you may want to make it the same type. In practice, we haven't found this to be an issue.
+1. Der Algorithmus versucht nicht, Teilbäume verschiedener Komponententypen abzugleichen. Wenn du feststellst, dass du zwischen zwei Komponententypen mit sehr ähnlicher Ausgabe wechseln, solltest du möglicherweise denselben Typ festlegen. In der Praxis ist dies normalerweise kein Problem.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
1. Der Algorithmus versucht nicht, Teilbäume verschiedener Komponententypen abzugleichen. Wenn du feststellst, dass du zwischen zwei Komponententypen mit sehr ähnlicher Ausgabe wechselst, solltest du möglicherweise denselben Typ festlegen. In der Praxis ist dies normalerweise kein Problem.
```

File: content/docs/reconciliation.md [hunk 10]
```diff
@@ -128,30 +128,30 @@ In order to solve this issue, React supports a `key` attribute. When children ha
...
 
-2. Keys should be stable, predictable, and unique. Unstable keys (like those produced by `Math.random()`) will cause many component instances and DOM nodes to be unnecessarily recreated, which can cause performance degradation and lost state in child components.
+2. Schlüssel sollten stabil, vorhersehbar und eindeutig sein. Instabile Schlüssel (wie die von `Math.random()` erzeugten) führen dazu, dass viele Komponenteninstanzen und DOM-Knoten unnötigerweise neu erstellt werden, was zu Leistungseinbußen und zum Verlust des Status in untergeordneten Komponenten führen kann.
```
Comment (miku86) [inline content/docs/reconciliation.md]:
```suggestion
2. Schlüssel sollten stabil, vorhersehbar und eindeutig sein. Instabile Schlüssel (wie die von `Math.random()` erzeugten) führen dazu, dass viele Komponenteninstanzen und DOM-Knoten unnötigerweise neu erstellt werden, was zu Leistungseinbußen und zum Verlust des States in untergeordneten Komponenten führen kann.
```

File: content/docs/reconciliation.md (hunk 6 above)
Comment (codejet) [reply content/docs/reconciliation.md]: yeah, i should probably revert it to `xml`. i did this without further reflecting on it...

File: content/docs/reconciliation.md (hunk 3 above)
Comment (codejet) [reply content/docs/reconciliation.md]: Yes, but it doesn't say "these", which would translate to "diese". But I still agree that your suggestion sounds better 👍

File: content/docs/reconciliation.md (hunk 2 above)
Comment (codejet) [reply content/docs/reconciliation.md]: Agreed

File: content/docs/reconciliation.md (hunk 5 above)
Comment (codejet) [reply content/docs/reconciliation.md]: both are fine i guess

**Comment by codejet** [comment]
@miku86 thanks a lot for the feedback. as no one else chimed in i implemented all the things you suggested.

**Pull request merged**